checks.d/mapreduce.py

```diff
@@ -0,0 +1,449 @@
+
+    def _set_metrics_from_json(self, tags, metrics_json, metrics):
+        '''
+        Parse the JSON response and set the metrics
+        '''
+        for status, (metric_name, metric_type) in metrics.iteritems():
+            metric_status = metrics_json.get(status)
+
+            if metric_status is not None:
+                self._set_metric(metric_name,
+                    metric_type,
+                    metric_status,
+                    tags)
+
+    def _set_metric(self, metric_name, metric_type, value, tags=None, device_name=None):
+        '''
+        Set a metric
+        '''
+        if metric_type == GAUGE:
+            self.gauge(metric_name, value, tags=tags, device_name=device_name)
+        elif metric_type == HISTOGRAM:
+            self.histogram(metric_name, value, tags=tags, device_name=device_name)
+        else:
+            self.log.error('Metric type "%s" unknown' % (metric_type))
+
+    def _rest_request_to_json(self, address, object_path, service_name=SERVICE_CHECK, *args, **kwargs):
+        '''
+        Query the given URL and return the JSON response
+        '''
+        response_json = None
+
+        service_check_tags = ['instance:%s' % self.hostname]
+
+        url = address
+
+        if object_path:
+            url = self._join_url_dir(url, object_path)
+
+        # Add args to the url
+        if args:
+            for directory in args:
+                url = self._join_url_dir(url, directory)
+
+        self.log.debug('Attempting to connect to "%s"' % url)
+
+        # Add kwargs as arguments
+        if kwargs:
+            query = '&'.join(['{0}={1}'.format(key, value) for key, value in kwargs.iteritems()])
+            url = urljoin(url, '?' + query)
+
+        try:
+            response = requests.get(url)
+            response.raise_for_status()
+            response_json = response.json()
+
+        except Timeout as e:
+            self.service_check(service_name,
+                AgentCheck.CRITICAL,
+                tags=service_check_tags,
+                message="Request timeout: {0}, {1}".format(url, e))
+            raise
+
+        except (HTTPError,
+                InvalidURL,
+                ConnectionError) as e:
+            self.service_check(service_name,
+                AgentCheck.CRITICAL,
+                tags=service_check_tags,
+                message="Request failed: {0}, {1}".format(url, e))
+            raise
+
+        except JSONDecodeError as e:
+            self.service_check(service_name,
+                AgentCheck.CRITICAL,
+                tags=service_check_tags,
+                message='JSON Parse failed: {0}, {1}'.format(url, e))
+            raise
+
+        except ValueError as e:
+            self.service_check(service_name,
+                AgentCheck.CRITICAL,
+                tags=service_check_tags,
+                message=e)
+            raise
+
+        else:
+            self.service_check(service_name,
```

Comment:
Instead of sending an `OK` service check every time the request is successful, we should probably only send the `OK` status once at the end of a check run on an instance (so at the end the `check` method for instance).